app/models/taxon_scoring_model.rb

```diff
@@ -0,0 +1,96 @@
+# Scoring Model for Taxon
+# the scoring model is represented in json with a tree structure where a parent will apply an operator to all the children and the leaf node could be a constant(const) or an attribute from the taxon (attr)
+# Currently the operators supported are addition (+), product (-) for nodes with multiple children and atomic transformation function for nodes with one children.
+# Here is an example of how the following scoring function can be translated to a json
+#
+# aggregate_score = |genus.NT.zscore|*species.NT.zscore*species.NT.rpm + |genus.NR.zscore|*species.NR.zscore*species.NR.rpm
+# {
+#  "op" : "+",
+#  "on" : [
+#    {
+#      "op": "*",
+#      "on": [
+#        {
+#          "op": "abs",
+#          "on": { "op": "attr", "on": "genus.NT.zscore" }
+#        },
+#        { "op": "attr", "on": "species.NT.zscore" },
+#        { "op": "attr", "on": "species.NT.rpm" }
+#      ]
+#    },
+#    {
+#      "op": "*",
+#      "on": [
+#        {
+#          "op": "abs",
+#          "on": { "op": "attr", "on": "genus.NR.zscore" }
+#        },
+#        { "op": "attr", "on": "species.NR.zscore" },
+#        { "op": "attr", "on": "species.NR.rpm" }
+#      ]
+#    }
+#  ]
+# }
+class TaxonScoringModel < ApplicationRecord
```

Comment:
Is this tree model generic enough? I see this as an unnecessary complexity. 

Why not just have a ruby library of scoring functions and store the name of the scoring function?
Seems simpler to me and provide way more flexibility.

Reply:
yes. it's very generic. it's a general representation of any arithmatic formula

Reply:
i would imagine people can plug in their function down the road and we can just store their formula in the database. also, katrina's model probably will change all the time, i don't want to change the code every time she (or anyone) updates the model.

Reply:
Some random thoughts or suggestions, mostly just to see what you think about those, not really pushing to get you to change this.

1) Can we store a lisp function in the database instead of a json abstract syntax tree?   Then we can use https://github.com/maryrosecook/littlelisp to run the stored lisp function in the browser or https://github.com/mooreryan/lisby to run the same lisp function in Ruby.

2) I am afraid to ask about performance here, but I see several approaches that could work to improve performance:
     a) The ruby backend could generate javascript that computes the formula, then ship that javascript to the browser, where it would get compiled and executed fast.
     b) Experimenting with novel formulas would be expected to run slow, but once the formulas are vetted and blessed, we can add ruby/js code for them, so they run natively and bypass this AST interpreter.

Overall I am slightly concerned that if we reinvent a lot of things then we might incur some maintenance cost -- but on the other hand if this lets you move faster and experiment faster in the product, we can have it this way in the short term and eventually replace with a more standard solution.

Reply:
Oh btw if we go the way of custom formulas, we should also do custom filters, all the way :)

Reply:
FWIW, interana had its own similar interpreter for query filters represented in a custom format (i think that was reverse polish and not json ast, but same difference);  That interpreter was written in C++, but eventually had to be replaced with on-demand LLVM compiler so the filters could run natively (much faster).  Here we have a lot less data to run this on, up to a few million rows perhaps?   So if this AST evaluator was implemented in C++, I wouldn't raise a  perf concern.   Only because ruby is already slow to begin with, before we ask it to eval a tree, am I slightly concerned --- but not overly concerned since this may still not be the slowest bit.  Just worth noting/measuring perhaps.

Reply:
Up to you...
But I would not call it generic, if it constrains you to arithmetic (eventually logic) functions. And in this case introduces unnecessary complexity to constrain what you can use. 
I guess it is for making customization of scoring more accessible to users...

If you want to follow this path anyway, I also agree with Boris that we should not reinvent the wheel: there are solutions out there with multi operations and multi language support, e.g. http://jsonlogic.com/

Also, could this be ran as part of the pipeline, as a post processing step? If it is run on demand as part of the web, every bit counts.

Reply:
This thing has to be done on the server side. (There might be some privacy/security issue of exposing it on the client side.) And we can't do it on the pipeline side, because it changes dynamically. 

We have to do it on the web part. I did a similar implementation in C++ before for delopying (generalized) linear models so I think this will be more than sufficient after adding a bit more features ( '/', identity functions etc). It's generic in a way that we dont have to change code every single time people update the model.  Sure if we were to do tree ensembles or deep learning models, we will have to do something else but for now and at our stage, i believe this is more than enough.

Reply:
exposing it : it means model coefficients.

Reply:
Changes dynamically because of the background model?
We can compute that too...